Exotic methods support

Sytten opened the pull request:
- Add support for exotic methods
- Add a macro to more easily implement those
    - it's a bit ugly but it works, I will improve it over time
    - Unsure about the get,set,has,delete names
- Remove send bound for async_with when not in parallel mode

Finishes #497

Commit:
add support for exotic objects

Commit:
add set_property

Commit:
add has/delete property

Commit:
cargo fmt

Commit:
add iterator test case to exotic

Commit:
fix

Sytten committed:
Add holder for exotic methods ffi

Sytten committed:
Warn user when trying to use CALLABLE and EXOTIC at the same time

Sytten committed:
Remove unused obj

We already have this

Sytten committed:
Add exotic macro

Sytten committed:
Add test and example for macro exotic

Sytten committed:
Fix example

Sytten committed:
Fix example 2

richarddd reviewed (approved):
Minor suggestions

DelSkayn commented on core/src/class.rs:
```diff
@@ -30,6 +31,9 @@ pub trait JsClass<'js>: Trace<'js> + JsLifetime<'js> + Sized {
 
+    /// Is this class exotic (e.g. will exotic_* methods be called with it)
+    const EXOTIC: bool = false;
+
```
Since a class can't be both exotic and callable, it is probably better to encode that in the type system, i.e. change these bools to something like `const KIND: ClassKind = ClassKind::Plain` where `ClassKind` is something like:
```rust
enum ClassKind{
    Plain,
    Callable,
    Exotic,
}
```

Sytten replied:
Right that is a good idea. Though I dont like the whole system of a single class ID for all rust classes.

DelSkayn reviewed (commented):
Two suggestions.

Sytten commented on core/src/result.rs:
```diff
@@ -689,6 +700,20 @@ impl<'js> Ctx<'js> {
+                qjs::JS_Throw(self.as_ptr(), qjs::JS_MKVAL(qjs::JS_TAG_EXCEPTION, 0));
+                -1
+            },
+        }
```
Kinda of in a crunch right now, I need to refactor the macro and the bool into an enum. If you want to help be my guess

xaptronic commented:
hey, came across this while optimizing Rust/JS interop patterns. will this PR include support for get_own_property and get_own_property_names? I noticed they're TODO in ExoticMethodsHolder — those are the main callbacks we'd need. happy to help with the remaining work if useful.

Sytten commented:
Honestly I dont have time to work on that right now, so @richarddd if you want to take it over or have someone else take it off my plate it works for me.

richarddd committed:
Address PR review comments

- Remove unused for<'a> lifetime bounds on exotic function type aliases
- Replace CALLABLE/EXOTIC bool consts with ClassKind enum
- Deduplicate handle_panic and handle_panic_exotic into shared inner method
- Remove unused ToString import

richarddd committed:
CI and cleanup

xaptronic commented:
NP - are these `get_own_property` and `get_own_property_names` sensible / possible to do as part of this? I noticed they also don't exist on the proxy objects - would this be more sensible to do in a follow up on both?

richarddd committed:
ci

richarddd commented:
> NP - are these `get_own_property` and `get_own_property_names` sensible / possible to do as part of this? I noticed they also don't exist on the proxy objects - would this be more sensible to do in a follow up on both?

@xaptronic seems in scope. will add!

richarddd committed:
Add get_own_property, get_own_property_names exotic methods

Sytten commented:
The reason I didnt do it previously is that the vtable is starting to get pretty big. I am not as fan that all classes use the same quickjs class id since it makes not nice to extend a rust class in js.

richarddd committed:
fix race

richarddd committed:
Merge branch 'master' into ef-exotic-class
